### PR TITLE
Replace Selenium with Playwright for browser tests

### PR DESCRIPTION
**Pull request by djamg**

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -59,6 +59,11 @@ def firefox_options(firefox_options):
     return firefox_options
 
 
+@pytest.fixture(scope="session")
+def browser_context_args(browser_context_args):
+    return {**browser_context_args, "ignore_https_errors": True}
```
Comment (jace) [inline tests/conftest.py:64]: Why is this needed?

Comment (djamg) [reply tests/conftest.py:64]: ignores SSL errors and certificate warnings

File: tests/e2e/account/pw_register_test.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+def check_recaptcha_loaded(page: Page) -> None:
+    page.wait_for_selector(
+        "#form-passwordlogin > div.g-recaptcha > div > div.grecaptcha-logo > iframe",
+        timeout=10000,
+    )
+
+
+@given("Anonymous visitor is on the home page")
+def given_anonuser_home_page(live_server, page) -> None:
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:25]: Specify type for Page here too, and in all other tests.

File: tests/e2e/account/pw_register_test.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    page.goto(live_server.url)
+    expect(page).to_have_title(re.compile("Test Hasgeek"))
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:27]: There's no point calling re.compile at runtime. If this function accepts a plain string, you can pass one, as this is a plain text pattern.

File: tests/e2e/account/pw_register_test.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+@when(
+    parsers.re(
+        "they navigate to the login page and submit"
+        " (?P<phone_or_email>a phone number|an email address)"
+    ),
+    target_fixture='anon_username',
+)
+def when_anonuser_navigates_login_and_submits(
+    app, live_server, phone_or_email, page
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:38]: Type `phone_or_email: str`

File: tests/e2e/account/pw_register_test.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+) -> dict[str, str]:
+    if phone_or_email == "a phone number":
+        username = '8123456789'
+    elif phone_or_email == "an email address":
+        username = 'anon@example.com'
+    else:
+        pytest.fail("Unknown username type")
+    page.click(".header__button")
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:46]: These should use single quotes as they're symbols.

File: tests/e2e/account/pw_register_test.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    check_recaptcha_loaded(page)
+    page.wait_for_selector("input[name=username]").fill(username)
+    page.click("#form-passwordlogin button")
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:49]: Likewise in both these lines – single quotes.

File: tests/e2e/account/pw_register_test.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+def given_twoflower_visits_project(
+    live_server, page, db_session, user_twoflower, new_project
+) -> None:
+    user_twoflower.password = TWOFLOWER_PASSWORD
+    user_twoflower.add_phone(TWOFLOWER_PHONE)
+    user_twoflower.add_email(TWOFLOWER_EMAIL)
+    new_project.publish()
+    db_session.add(new_project)
+    db_session.commit()
+    page.goto(live_server.url + new_project.account.urlname + '/' + new_project.name)
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:165]: Use `new_project.absolute_url` instead of constructing it like this.

**Review by jace — CHANGES_REQUESTED** [review]
Many strings with double quotes (esp CSS class selectors) should be in single quotes to indicate they're symbols – any change to the value here should be synced with the same string elsewhere. Other issues are as indicated in comments.

File: tests/e2e/account/pw_register_test.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+scenarios('account/register.feature')
+pytestmark = pytest.mark.usefixtures('live_server')
+
+TWOFLOWER_EMAIL = 'twoflower@example.org'
+TWOFLOWER_PHONE = '+12015550123'
+TWOFLOWER_PASSWORD = 'te@pwd3289'  # nosec
+
+
+def check_recaptcha_loaded(page: Page) -> None:
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:15]: This should be `wait_until_recaptcha_loaded` and it should return a bool indicating success or failure.

File: tests/e2e/account/pw_register_test.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    return {'phone_or_email': phone_or_email, 'username': username}
+
+
+@then("they are prompted for their name and the OTP, which they provide")
+def then_anonuser_prompted_name_and_otp(live_server, anon_username, page: Page) -> None:
+    page.wait_for_selector('input[name=fullname]').fill('Twoflower')
+    if anon_username['phone_or_email'] == "a phone number":
+        otp = live_server.transport_calls.sms[-1].vars['otp']
+    elif anon_username['phone_or_email'] == "an email address":
+        otp = live_server.transport_calls.email[-1].subject.split(' ')[-1]
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:57]: For safety, this should use a regex pattern to extract the OTP from the subject. If the subject format changes, this test will fail for regex not matching instead of incorrect OTP – making the problem easier to spot.

File: tests/e2e/account/pw_register_test.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        pytest.fail("Unknown username type")
+    page.wait_for_selector('input[name=otp]').fill(otp)
+    page.click('#form-otp button')
+
+
+@then("they get an account and are logged in")
+def then_they_are_logged_in(live_server, page: Page) -> None:
+    assert (
+        page.get_by_text("You are now one of us. Welcome aboard!").inner_text()
+        == "You are now one of us. Welcome aboard!"
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:68]: Don't repeat the string. You can instead test that the selector found a match.

File: tests/e2e/account/pw_register_test.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+@given("Twoflower visitor is on the home page")
+def when_twoflower_visits_homepage(
+    live_server, page: Page, db_session, user_twoflower
+) -> None:
+    page.goto(live_server.url)
+    user_twoflower.password = TWOFLOWER_PASSWORD
+    user_twoflower.add_phone(TWOFLOWER_PHONE)
+    user_twoflower.add_email(TWOFLOWER_EMAIL)
+    db_session.commit()
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:80]: IMO, move these lines into a fixture, something like `user_twoflower_with_password_and_contacts`. The fixture depends on the underlying fixtures and returns the user object.

File: tests/e2e/account/pw_register_test.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    return {'phone_or_email': phone_or_email, 'username': username}
+
+
+@given("Twoflower is on the project page")
+def given_twoflower_visits_project(
+    live_server, page: Page, db_session, user_twoflower, new_project
+) -> None:
+    user_twoflower.password = TWOFLOWER_PASSWORD
+    user_twoflower.add_phone(TWOFLOWER_PHONE)
+    user_twoflower.add_email(TWOFLOWER_EMAIL)
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:161]: These lines should use the same fixture as above.

File: tests/e2e/account/pw_register_test.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    new_project.publish()
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:162]: You've got this twice. Move this into a `published_project` fixture.

File: tests/e2e/account/pw_register_test.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+@given("the server uses Recaptcha")
+def given_server_uses_recaptcha(
+    live_server, db_session, user_twoflower, new_project, funnel
+) -> None:
+    user_twoflower.password = TWOFLOWER_PASSWORD
+    user_twoflower.add_phone(TWOFLOWER_PHONE)
+    user_twoflower.add_email(TWOFLOWER_EMAIL)
+    new_project.publish()
+    db_session.add(new_project)
+    db_session.commit()
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:177]: One again, two reusable fixtures.

**Review by jace — CHANGES_REQUESTED** [review]
There's a lot of repetition that should be moved into fixtures. One OTP extraction needs regex instead of the string.

File: tests/e2e/account/pw_register_test.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+pytestmark = pytest.mark.filterwarnings(
+    "ignore:Object of type <AccountPhone> not in session",
+    "ignore:Object of type <AccountEmail> not in session",
+)
+
+
+@pytest.fixture()
+def published_project(db_session, new_project):
+    new_project.publish()
+    db_session.add(new_project)
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:26]: This is not needed. It's already in the session.

File: tests/e2e/account/pw_register_test.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+scenarios('account/register.feature')
+pytestmark = pytest.mark.usefixtures('live_server')
+
+TWOFLOWER_EMAIL = 'twoflower@example.org'
+TWOFLOWER_PHONE = '+12015550123'
+TWOFLOWER_PASSWORD = 'te@pwd3289'  # nosec
+
+pytestmark = pytest.mark.filterwarnings(
+    "ignore:Object of type <AccountPhone> not in session",
+    "ignore:Object of type <AccountEmail> not in session",
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:19]: Don't filter these. The problem is in the `add` classmethod of EmailAddress and PhoneNumber. It should be resolved there.

File: tests/e2e/account/pw_register_test.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+)
+
+
+@pytest.fixture()
+def published_project(db_session, new_project):
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:24]: Specify types: `new_project: models.Project) -> models.Project:`

File: tests/e2e/account/pw_register_test.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+    new_project.publish()
+    db_session.add(new_project)
+    db_session.commit()
+    return new_project
+
+
+@pytest.fixture()
+def user_twoflower_with_password_and_contact(db_session, user_twoflower):
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:32]: Also needs type definitions here.

File: tests/e2e/account/pw_register_test.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+    user_twoflower.password = TWOFLOWER_PASSWORD
+    user_twoflower.add_phone(TWOFLOWER_PHONE)
+    user_twoflower.add_email(TWOFLOWER_EMAIL)
+    db_session.add(user_twoflower)
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:36]: This line is not required.

File: tests/e2e/account/pw_register_test.py [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+    ),
+    target_fixture='anon_username',
+)
+def when_anonuser_navigates_login_and_submits(
+    app, live_server, phone_or_email: str, page: Page
+) -> dict[str, str]:
+    if phone_or_email == "a phone number":
+        username = '8123456789'
+    elif phone_or_email == "an email address":
+        username = 'anon@example.com'
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:67]: I think you should put both these constants at the top of the file. We'll have to audit for where we're using these values across teats and put them into shared fixtures – apart from ensuring we're not using real data.

File: tests/e2e/account/pw_register_test.py [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+
+@then("they are prompted for their name and the OTP, which they provide")
+def then_anonuser_prompted_name_and_otp(live_server, anon_username, page: Page) -> None:
+    page.wait_for_selector('input[name=fullname]').fill('Twoflower')
+    if anon_username['phone_or_email'] == "a phone number":
+        otp = live_server.transport_calls.sms[-1].vars['otp']
+    elif anon_username['phone_or_email'] == "an email address":
+        subject = live_server.transport_calls.email[-1].subject
+        otp_pattern = r"\b\d{4}\b"
+        otp = re.search(otp_pattern, subject).group(0)
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:85]: The var is not needed. Specify the pattern directly here – but with single quotes as it's not a UI string.

File: tests/e2e/account/pw_register_test.py [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+    page.click('#form-otp button')
+
+
+@then("they get an account and are logged in")
+def then_they_are_logged_in(
+    live_server, user_twoflower_with_password_and_contact, page: Page
+) -> None:
+    assert (
+        page.wait_for_selector('.alert__text').inner_text()
+        == "You are now one of us. Welcome aboard!"
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:98]: This string is going to be a problem as it's UI and could change unexpectedly. It can stay for now, but we'll need a better marker (future PR).

File: tests/e2e/account/pw_register_test.py [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+    )
+
+
+@given("Twoflower visitor is on the home page")
+def when_twoflower_visits_homepage(
+    live_server, page: Page, db_session, user_twoflower_with_password_and_contact
+) -> None:
+    page.goto(live_server.url)
+    db_session.commit()
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:107]: Always do db operations first, browser navigation second.

File: tests/e2e/account/pw_register_test.py [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+@when("they click on follow")
+def when_they_click_follow(page: Page) -> None:
+    page.wait_for_selector("#register-nav").click()
+
+
+@then("a register modal appears")
+def then_register_modal_appear(page: Page) -> None:
+    assert (
+        page.wait_for_selector('xpath=//*[@id="passwordform"]/p[2]').inner_text()
+        == "Tell us where you’d like to get updates. We’ll send an OTP to confirm."
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:158]: Another UI string to change for the next PR.

File: tests/e2e/account/pw_register_test.py [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+@when(
+    parsers.re("they enter (?P<phone_or_email>a phone number|an email address)"),
+    target_fixture='anon_username',
+)
+def when_they_enter_email(page: Page, phone_or_email: str) -> dict[str, str]:
+    wait_until_recaptcha_loaded(page)
+    if phone_or_email == "a phone number":
+        username = '8123456789'
+    elif phone_or_email == "an email address":
+        username = 'anon@example.com'
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:171]: Two more constants.

File: tests/e2e/account/pw_register_test.py [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+    page.wait_for_selector('input[name=username]').fill(username)
+    page.click('#form-passwordlogin button')
+    return {'phone_or_email': phone_or_email, 'username': username}
+
+
+@given("Twoflower is on the project page")
+def given_twoflower_visits_project(
+    live_server,
+    page: Page,
+    db_session,
```
Comment (jace) [inline tests/e2e/account/pw_register_test.py:183]: This doesn't strictly matter, but db_session should be first, then database fixtures, then the server, then the browser.

I remember there was an issue with fixture order in one pytest-bdd test. Reordering fixed it.

**Review by jace — CHANGES_REQUESTED** [review]
Some cosmetic changes left. You should also remove the pytest-selenium tests, remove the dependency, remove the pin on selenium as required by pytest-selenium, and also remove the ignore rule for it in pip-audit. Also remove the `playright` prefix on the file names – this PR should completely switch over from pytest-selenium to playwright.

**Review by jace — APPROVED** [review]
